File: 3.Longest_Substring/substringSet.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;
class Solution{
    public:
        int lengthOfLongestSubstring(string s){
            int n = s.size();
            int ans = 0;
            unordered_set<char> index;
            int left = 0;

            for(int right=0; right<n; right++){
                if(index.count(s[right])== 0){
                    index.insert(s[right]);
                    ans = max(ans, right - left + 1);
                } else{
                    while(index.count(s[right]) != 0){
                        index.erase(s[left]);
                        left++;
                    }
                    index.insert(s[right]);
                }
            }
            return ans;
        }
};

int main(){
    string s = "abcabcbb";
    Solution sol;
    cout << sol.lengthOfLongestSubstring(s) << endl;
    return 0;
}

File: 3.Longest_Substring/substringSet.java
import java.util.*;
import java.util.HashSet;
import java.util.Set;


class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        Set<Character> charSet = new HashSet<>();
        int left = 0;
        
        for (int right = 0; right < n; right++) {
            if (!charSet.contains(s.charAt(right))) {
                charSet.add(s.charAt(right));
                maxLength = Math.max(maxLength, right - left + 1);
            } else {
                while (charSet.contains(s.charAt(right))) {
                    charSet.remove(s.charAt(left));
                    left++;
                }
                charSet.add(s.charAt(right));
            }
        }
        
        return maxLength;
    }
}


public class substringSet{
    public static void main(String[] args){
        Solution sol = new Solution();
        String s = "abcabcbb";
        System.out.println(sol.lengthOfLongestSubstring(s));
    }
}